src/cli/migrate/nodeVersion/index.ts

```diff
@@ -6,131 +6,122 @@ import fs from 'fs-extra';
+  | { file: string; files?: never };
 
-const SHA_REGEX = /(?<=node.*)(@sha256:[a-f0-9]{64})/gm;
+type SubPatch = FileSelector & {
+  tests?: Array<() => Promise<boolean>>;
+  regex?: RegExp;
+  replace: string;
+};
 
-const subPatches: SubPatch[] = [
-  { id: 'nvmrc', file: '.nvmrc', replace: '<%- version %>\n' },
+const subPatches = ({
+  nodeVersion,
+  nodeTypesVersion,
+  ECMAScriptVersion,
+}: Versions): SubPatch[] => [
+  { file: '.nvmrc', replace: `${nodeVersion}\n` },
   {
-    id: 'Dockerfile-1',
     files: '**/Dockerfile*',
-    test: /^FROM(.*) (public.ecr.aws\/docker\/library\/)?node:[0-9.]+(@sha256:[a-f0-9]{64})?(\.[^- \n]+)?(-[^ \n]+)?( .+|)$/gm,
-    replace: 'FROM$1 $2node:<%- version %>$3$5$6',
+
+    regex:
+      /^FROM(.*) (public.ecr.aws\/docker\/library\/)?node:([0-9]+(?:\.[0-9]+(?:\.[0-9]+)?)?)(-[a-z0-9]+)?(@sha256:[a-f0-9]{64})?( .*)?$/gm,
+    replace: `FROM$1 $2node:${nodeVersion}$4$6`,
   },
   {
-    id: 'Dockerfile-2',
     files: '**/Dockerfile*',
-    test: /^FROM(.*) gcr.io\/distroless\/nodejs\d+-debian(.+)$/gm,
-    replace: 'FROM$1 gcr.io/distroless/nodejs<%- version %>-debian$2',
+    regex:
+      /^FROM(.*) gcr.io\/distroless\/nodejs\d+-debian(\d+)(@sha256:[a-f0-9]{64})?(\.[^- \n]+)?(-[^ \n]+)?( .+|)$/gm,
+    replace: `FROM$1 gcr.io/distroless/nodejs${nodeVersion}-debian$2$4$5$6`,
+  },
+
+  {
+    files: '**/serverless*.y*ml',
+    regex: /nodejs\d+.x/gm,
+    tests: [isPatchableServerlessVersion],
+    replace: `nodejs${nodeVersion}.x`,
   },
   {
-    id: 'serverless',
     files: '**/serverless*.y*ml',
-    test: /nodejs\d+.x/gm,
-    replace: 'nodejs<%- version %>.x',
+    regex: /node\d+/gm,
+    tests: [isPatchableServerlessVersion],
+    replace: `node${nodeVersion}`,
+  },
+
+  {
+    files: '**/infra/**/*.ts',
+    regex: /NODEJS_\d+_X/g,
+    replace: `NODEJS_${nodeVersion}_X`,
+  },
+  {
+    files: '**/infra/**/*.ts',
+    regex: /(target:\s*'node)(\d+)(.+)$/gm,
+    replace: `$1${nodeVersion}$3`,
   },
-  [
-    {
-      id: 'cdk-1',
-      files: '**/infra/**/*.ts',
-      test: /NODEJS_\d+_X/g,
-      replace: 'NODEJS_<%- version %>_X',
-    },
-    {
-      id: 'cdk-2',
-      files: '**/infra/**/*.ts',
-      test: /(target:\s*'node)(\d+)(.+)$/gm,
-      replace: '$1<%- version %>$3',
-    },
-  ],
+
   {
-    id: 'buildkite',
     files: '**/.buildkite/*',
-    test: /(image: )(public.ecr.aws\/docker\/library\/)?(node:)[0-9.]+(\.[^- \n]+)?(-[^ \n]+)?$/gm,
-    replace: '$1$2$3<%- version %>$5',
+    regex:
+      /(image: )(public.ecr.aws\/docker\/library\/)?(node:)[0-9.]+(\.[^- \n]+)?(-[^ \n]+)?$/gm,
+    replace: `$1$2$3${nodeVersion}$5`,
   },
   {
-    id: 'node-version',
     files: '.node-version*',
-    test: /(v)?\d+\.\d+\.\d+(.+)?/gm,
-    replace: '$1<%- version %>$2',
+    regex: /(v)?\d+\.\d+\.\d+(.+)?/gm,
+    replace: `$1${nodeVersion}$2`,
+  },
+
+  {
+    files: '**/package.json',
+    regex: /(\\?"@types\/node\\?": \\?")(\^)?[0-9.]+(\\?(",?)\\?n?)/gm,
```

Comment:
General question, do we need all these `\\?`s? The patterns seems to work without them but maybe I'm missing something obvious

Reply:
They exist because I am matching on the stringified version of the `package.json` 😢

Reply:
Hmm, are we double-encoding the `package.json`? If we can avoid that then we should be fine:

```javascript
> /"hello"/.exec('{"hello": "there"}')
[ '"hello"', index: 1, input: '{"hello": "there"}', groups: undefined ]
```

Reply:
It was my own tests that were fighting me... 😅